src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,472 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2021 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Checks that a local variable is declared and/or assigned, but not used.
+ * Doesn't check
+ * <a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.30.1">
+ * pattern variables</a>.
+ * </p>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;UnusedLocalVariable&quot;/&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * class Test {
+ *
+ *     int a;
+ *
+ *     {
+ *         int k = 12; // violation, assigned and updated but never used
+ *         k++;
+ *     }
+ *
+ *     Test(int a) { // ok as 'a' is a method parameter not a local variable
+ *         this.a = 12;
+ *     }
+ *
+ *     void method(int b) {
+ *         int a = 10; // violation
+ *         int[] arr = {1, 2, 3}; // violation
+ *         int[] anotherArr = {1}; // ok
+ *         anotherArr[0] = 4;
+ *     }
+ *
+ *     String convertValue(String newValue) {
+ *         String s = newValue.toLowerCase(); // violation
+ *         return newValue.toLowerCase();
+ *     }
+ *
+ *     void read() throws IOException {
+ *         BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+ *         String s; // violation
+ *         while ((s = reader.readLine()) != null) {
+ *         }
+ *         try (BufferedReader reader1 = new BufferedReader(new FileReader("abc.txt"))) { // ok
```

Comment:
please add explanation here, just an `// ok` comment leads me to believe that there exists a case where this could be a violation, maybe something like `resources are not checked` or `ok as reader1 is a resource`?

Reply:
done

Reply:
Resolved.